rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,96 @@
+to 1 read, although such efficiency is possible for special cases where input
+tensor is small enough.
+
+## Proposal
+Use the following formula to calculate variance:
+```
+Var(X) = E(X^2) - (E(X))^2
+```
+where X = input tensor, E = average
+
+This formula allows hardware to calculate both mean and variance in single pass
+over input tensor.
+For reference, the canonical formula is:
+```
+Var(X) = E((X - Mean(X))^2)
+```
+The benefit of new formula is 25-33% performance improvement. Batchnorm kernel
+is memory-bound, with new formula it will need to read input tensor 2 times 
+(1: mean&var, 2: normalize) instead of 3 times (1: mean, 2: var, 3: normalize)
+
+This formula will be used whenever batchnorm kernel calculates mean and 
+variance, which means forward propagation in training. However, with 
+non-standard combination of flags, oneDNN's batchnorm implementaton may be
+configured to calculate those two values in inference as well.
+
+## Design
+Algorithm for calculating variance will be implemented as:
+```
+Sum = 0
+SumOfSquares = 0
+for each x in X:
+	Sum += x
+	SumOfSquares += (x*x)
+Variance = SumOfSquares/N - (Sum/N)^2
+Variance = max(0, Variance)
+```
+where N = number of items in given channel in tensor.
+
+In the formula there's a subtraction of two large values which is vulnerable
+to catastrophic cancellation. Rounding errors in the two big numbers may be
+large in relation to result of the formula. In order to minimize rounding
+errors:
+- Sum and SumOfSquares will be stored in fp32 precision, regardless of input
+	data type
+- Kahan summation algorithm will be used
+
+The last line, max( ) is there to make sure rounding errors won't cause
+variance to go negative. Further in batchnorm calculations square root is taken
+from variance. Negative value would result in NaN.
+
+New formula is considered experimental for now. By default the old one will be
+used. There needs to be a way for user to choose formula.
+
+### Enable through build flag and environmental variable
+A new build flag `DNNL_EXPERIMENTAL` is introduced, disabled by default.
+When built with this flag enabled, oneDNN will read env var 
+`DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS` and will select bnorm formula 
+according to that flag. Default is the slower but numerically stable formula.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+
+Cons:
+- No fine-tuning per instance of bnorm primitive
+- Harder to reproduce bugs related to bnorm: value of this flag will need to be
+ reported in bug reports to allow reproduction.
+ - Inconvenient long-term. State of oneDNN will have a dependency which is not
+ obvious to users.
+
+## Layernorm
+Batchnorm and layernorm primitives work according to the same principles and
+formulas. Only difference is in dimension across which the normalization is
+performed. Therefore layernorm's performance may benefit from using the new
+formula for calculating mean & variance.
```

Comment:
Are we sure this is useful for layermorm? In that case, statistics are computed across channels, and this is typically small enough so that the two pass algorithm should happen in register/slm.

Reply:
It depends on tensor size. Indeed, with BERT's layernorms we encountered so far we don't need it.